src/sparsebase/utils/io/reader.cc

```diff
@@ -125,14 +126,223 @@ EdgeListReader<IDType, NNZType, ValueType>::~EdgeListReader(){};
 
 template <typename IDType, typename NNZType, typename ValueType>
 MTXReader<IDType, NNZType, ValueType>::MTXReader(std::string filename,
-                                                 bool weighted,
                                                  bool convert_to_zero_index)
-    : filename_(filename), weighted_(weighted),
-      convert_to_zero_index_(convert_to_zero_index) {}
+    : filename_(filename),
+      convert_to_zero_index_(convert_to_zero_index) {
+  std::ifstream fin(filename_);
+
+  if (fin.is_open()) {
+    std::string header_line;
+    std::getline(fin, header_line);
+    // parse first line
+    options_ = ParseHeader(header_line);
+  } else {
+    throw ReaderException("Wrong matrix market file name\n");
+  }
+}
+
+template <typename IDType, typename NNZType, typename ValueType>
+typename MTXReader<IDType, NNZType, ValueType>::MTXOptions
+MTXReader<IDType, NNZType, ValueType>::ParseHeader(
+    std::string header_line) const {
+  std::stringstream line_ss(header_line);
+  MTXOptions options;
+  std::string prefix, object, format, field, symmetry;
+  line_ss >> prefix >> object >> format >> field >> symmetry;
+  if (prefix != MMX_PREFIX)
+    throw ReaderException("Wrong prefix in a matrix market file");
+  // parsing Object option
+  if (object == "matrix") {
+    options.object =
+        MTXReader<IDType, NNZType, ValueType>::MTXObjectOptions::matrix;
+  } else if (object == "vector") {
+    options.object =
+        MTXReader<IDType, NNZType, ValueType>::MTXObjectOptions::matrix;
+    throw ReaderException("Matrix market reader does not currently support reading vectors.");
+  } else {
+    throw ReaderException(
+        "Illegal value for the 'object' option in matrix market header");
+  }
+  // parsing format option
+  if (format == "array") {
+    options.format =
+        MTXReader<IDType, NNZType, ValueType>::MTXFormatOptions::array;
+  } else if (format == "coordinate") {
+    options.format =
+        MTXReader<IDType, NNZType, ValueType>::MTXFormatOptions::coordinate;
+  } else {
+    throw ReaderException(
+        "Illegal value for the 'format' option in matrix market header");
+  }
+  // parsing field option
+  if (field == "real") {
+    options.field =
+        MTXReader<IDType, NNZType, ValueType>::MTXFieldOptions::real;
+    if constexpr(std::is_same<void, ValueType>::value) throw ReaderException("You are reading the values of the matrix market file into a void array");
```

Comment:
Just an idea since this exception appears and will appear in other readers: what if we created a class for this exception? VoidValueArrayException ?

Reply:
Interesting idea, but I think this is informative enough unless we can add more info through that other exception class. Like this exception tells the user what they want to know, that it's in the readers, that it's matrix market related, and the message. Do you think the other exception type would be more informative or more generalizable?